src/audio/asrc/asrc.c

```diff
@@ -0,0 +1,615 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+
+#include <sof/audio/asrc/asrc_farrow.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/cache.h>
+#include <sof/math/numbers.h>
+#include <sof/trace/trace.h>
+#include <sof/common.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define trace_asrc(__e, ...)					\
+	trace_event(TRACE_CLASS_ASRC, __e, ##__VA_ARGS__)
+#define tracev_asrc(__e, ...)					\
+	tracev_event(TRACE_CLASS_ASRC, __e, ##__VA_ARGS__)
+#define trace_asrc_error(__e, ...)				\
+	trace_error(TRACE_CLASS_ASRC, __e, ##__VA_ARGS__)
+
+/* asrc component private data */
+struct comp_data {
+	uint32_t sink_rate;
+	uint32_t source_rate;
+	uint32_t sink_format;
+	uint32_t source_format;
+	uint32_t src_size;
+	uint32_t buf_size;
+	int32_t drift;
+	void *buf;
+	void *ibuf[PLATFORM_MAX_CHANNELS];
+	void *obuf[PLATFORM_MAX_CHANNELS];
+	int frames;
+	int source_frames;
+	int sink_frames;
+	int source_frames_max;
+	int sink_frames_max;
+	int data_shift;
+	struct asrc_farrow *src_obj;
+	enum asrc_operation_mode mode;
+	void (*asrc_func)(struct comp_dev *dev,
+			  struct comp_buffer *source,
+			  struct comp_buffer *sink,
+			  int *consumed,
+			  int *produced);
+};
+
+/* In-line functions */
+
+static inline void src_inc_wrap(int32_t **ptr, int32_t *end, size_t size)
+{
+	if (*ptr >= end)
+		*ptr = (int32_t *)((size_t)*ptr - size);
```

Comment:
@plbossart (sorry for spamming on a different PR) Should I use uintptr_t instead of char * in the places where I added casts in my own PR #2023?

Reply:
I'll propose to next version this kind of casts. Is it OK?

`*ptr = (int32_t *)((uint8_t *)*ptr - size);
`

Reply:
@singalsu I believe simply doing `(char *)` instead of `(uint8_t *)` is better because uint8_t may be a char on our platforms but in case we move to a really weird one it may become funny.

Reply:
@paulstelian97 you should be able to typedef uint8_t * to char * on your weird platform. I do remeber ARM had some unsigned char weirdness a while back but not sure if this is what you are referring to.

Reply:
Oh on ARM it's fine as it's still of size 1 so pointer arithmetic is fine. But given our consistent variant of doing it with my recent change with the void pointer arithmetic I'd say having two different ways across the same project isn't exactly great. And I sure don't want to go through all of them again, especially since this time compiler errors won't help me anymore.

Reply:
Oh, I thought we had removed all the void * arith now ? Is there still some pending removal ?

Reply:
I need to make a PR to fix SRC. There's a similar mistake with cast to size_t.

Reply:
@lgirdwood The actual void * arithmetic was removed, at least as much as the compiler forced me to do it once I added said compile flag. It's just that I used `char *` for all the casts to do that.

Reply:
Got approve for my SRC PR with similar fix so assume this is OK for ASRC too!